includes/polling/storage/hrstorage.inc.php

```diff
@@ -8,6 +8,31 @@
 
 $entry = $storage_cache['hrstorage'][$storage['storage_index']];
 
+/*
+ * It is possible that the HOST-RESOURCES-MIB::hrStorageTable has updated between
+ * discovery and polling.    If we assume it hasn't we can end up assigning the
+ * values of the wrong mount point to another mount point.   This can cause
+ * issues with not only display, but also alarming.
+ *
+ * We can compare the description expected with one from the hrStorageIndex entry
+ * and if they are not equal then we can filter the table to return an entry
+ * that does match.
+ *
+ * (example of how to debug --> ./poller.php -h 42 -r -f -m storage -d)
+ *
+ */
+
+if( $entry['hrStorageDescr'] !== $descr ){
```

Comment:
If the data is already fetched, why not always updated it?

Reply:
I'm not sure.   

All the data is fetched so in theory a 'rediscovery' could be done each time to update the appropriate indexes etc in the database.

I assume it was done this way originally because
* it is lightweight and fast
* storage indexes on servers didn't change much

Which is probably a good set of assumptions, and this just adds a little check incase something has changed between discovery and polling (like temporary devices, user-login /tmpfs stuff, USBs, etc... ) that has caused the hrStorageIndex to change

Reply:
It was originally done because the MIB specifies that indexes must not change.

As far as the MIB is concerned a removable disk is a new disk if it has been removed and re-plugged in.

Reply:
I can't see that spelt out clearly in the MIB although it does state hrStorageIndex is "A unique value for each logical storage area contained by the host." so you could consider that it should persist.

Looking at net-snmp's snmpd on a recent Ubuntu it does not persist between restarts of snmpd (although interestingly net-snmp does persist ifIndexes in /var/lib/snmpd).
 
One behavioural difference with my tested net-snmp implementation is that users filesystems in /run/user/{uid} all appear inserted with hrStorageIndex's past existing index's rather than getting stuffed into the middle like JunOS EVO's snmpd implementation does.

We will obviously kick the vendor, however I think the patch has some value as a sanity check for the messy world we live in (and all those JunOS EVO versions out there that will never get upgraded etc...)